File: src/pixiis/library/gog.py
"""GOG Galaxy library provider — detects games via Windows registry."""

from __future__ import annotations

import subprocess
import sys
import webbrowser
from pathlib import Path
from typing import TYPE_CHECKING

from pixiis.core.types import AppEntry, AppSource

if TYPE_CHECKING:
    from pixiis.core.config import Config

_GOG_REG_KEY = r"SOFTWARE\WOW6432Node\GOG.com\Games"


class GOGProvider:
    """Discover and launch GOG Galaxy games."""

    def __init__(self, config: Config) -> None:
        self._config = config

    @property
    def name(self) -> str:
        return "gog"

    def is_available(self) -> bool:
        if sys.platform != "win32":
            return False
        try:
            import winreg

            key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _GOG_REG_KEY)
            winreg.CloseKey(key)
            return True
        except Exception:
            return False

    def scan(self) -> list[AppEntry]:
        if sys.platform != "win32":
            return []

        try:
            import winreg

            root = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _GOG_REG_KEY)
        except Exception:
            return []

        apps: list[AppEntry] = []
        try:
            index = 0
            while True:
                try:
                    subkey_name = winreg.EnumKey(root, index)
                    entry = self._read_game_key(root, subkey_name)
                    if entry is not None:
                        apps.append(entry)
                except OSError:
                    break
                index += 1
        finally:
            winreg.CloseKey(root)

        return apps

    def launch(self, app: AppEntry) -> None:
        game_id = app.metadata.get("game_id", "")
        if game_id:
            webbrowser.open(f"goggalaxy://openGameView/{game_id}")
        elif app.exe_path and app.exe_path.exists():
            cf = subprocess.CREATE_NO_WINDOW if sys.platform == "win32" else 0
            subprocess.Popen(
                [str(app.exe_path)],
                cwd=str(app.exe_path.parent),
                creationflags=cf,
            )

    def get_icon(self, app: AppEntry) -> Path | None:
        return None

    # -- internals -----------------------------------------------------------

    @staticmethod
    def _read_game_key(root, subkey_name: str) -> AppEntry | None:
        """Read a single game entry from the GOG registry."""
        try:
            import winreg

            key = winreg.OpenKey(root, subkey_name)
        except OSError:
            return None

        try:
            game_name = _reg_value(key, "gameName")
            game_id = _reg_value(key, "gameID")
            exe = _reg_value(key, "exe")
            install_path = _reg_value(key, "path")

            if not game_name or not game_id:
                return None

            exe_path: Path | None = None
            if exe:
                exe_path = Path(exe)
            elif install_path:
                exe_path = Path(install_path)

            return AppEntry(
                id=f"gog:{game_id}",
                name=game_name,
                source=AppSource.GOG,
                launch_command=f"goggalaxy://openGameView/{game_id}",
                exe_path=exe_path,
                metadata={"game_id": game_id},
            )
        finally:
            import winreg

            winreg.CloseKey(key)


def _reg_value(key, value_name: str) -> str:
    """Safely read a string value from a registry key."""
    try:
        import winreg

        val, _ = winreg.QueryValueEx(key, value_name)
        return str(val).strip()
    except Exception:
        return ""

File: src-tauri/src/library/gog.rs
//! GOG Galaxy library provider — port of `src/pixiis/library/gog.py`.
//!
//! Walks `HKLM\SOFTWARE\WOW6432Node\GOG.com\Games\<id>`, reads
//! `gameName`, `gameID`, `exe`, `path`, and produces one AppEntry per
//! installed game whose launch_command is the
//! `goggalaxy://openGameView/{game_id}` URL Galaxy already advertises.

use std::path::PathBuf;

use serde_json::{Map, Value};

use super::Provider;
use crate::types::{AppEntry, AppSource};

#[cfg(target_os = "windows")]
const GOG_REG_KEY: &str = r"SOFTWARE\WOW6432Node\GOG.com\Games";

pub struct GogProvider;

impl GogProvider {
    pub fn new() -> Self {
        Self
    }
}

impl Default for GogProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl Provider for GogProvider {
    fn name(&self) -> &'static str {
        "gog"
    }

    fn is_available(&self) -> bool {
        #[cfg(target_os = "windows")]
        {
            registry::root_exists()
        }
        #[cfg(not(target_os = "windows"))]
        {
            false
        }
    }

    fn scan(&self) -> Vec<AppEntry> {
        #[cfg(target_os = "windows")]
        {
            registry::scan_games()
                .into_iter()
                .filter_map(|f| build_entry(&f))
                .collect()
        }
        #[cfg(not(target_os = "windows"))]
        {
            Vec::new()
        }
    }
}

/// Raw fields read from a single GOG registry game key. Split out so the
/// AppEntry-building logic (and its tests) does not depend on the
/// Windows registry being present.
#[derive(Debug, Default, Clone)]
#[cfg_attr(not(target_os = "windows"), allow(dead_code))]
pub(crate) struct GogGameFields {
    pub game_name: String,
    pub game_id: String,
    pub exe: String,
    pub install_path: String,
}

#[cfg_attr(not(target_os = "windows"), allow(dead_code))]
pub(crate) fn build_entry(f: &GogGameFields) -> Option<AppEntry> {
    if f.game_name.trim().is_empty() || f.game_id.trim().is_empty() {
        return None;
    }

    let exe_path: Option<PathBuf> = if !f.exe.is_empty() {
        Some(PathBuf::from(&f.exe))
    } else if !f.install_path.is_empty() {
        Some(PathBuf::from(&f.install_path))
    } else {
        None
    };

    let mut metadata = Map::new();
    metadata.insert("game_id".into(), Value::String(f.game_id.clone()));

    Some(AppEntry {
        id: format!("gog:{}", f.game_id),
        name: f.game_name.trim().to_string(),
        source: AppSource::Gog,
        launch_command: format!("goggalaxy://openGameView/{}", f.game_id),
        exe_path,
        icon_path: None,
        art_url: None,
        metadata,
    })
}

#[cfg(target_os = "windows")]
mod registry {
    use super::{GogGameFields, GOG_REG_KEY};
    use winreg::enums::HKEY_LOCAL_MACHINE;
    use winreg::RegKey;

    pub fn root_exists() -> bool {
        let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
        hklm.open_subkey(GOG_REG_KEY).is_ok()
    }

    pub fn scan_games() -> Vec<GogGameFields> {
        let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
        let Ok(root) = hklm.open_subkey(GOG_REG_KEY) else {
            return Vec::new();
        };

        let mut out = Vec::new();
        for sub in root.enum_keys().flatten() {
            let Ok(key) = root.open_subkey(&sub) else {
                continue;
            };
            let fields = GogGameFields {
                game_name: read_string(&key, "gameName"),
                game_id: read_string(&key, "gameID"),
                exe: read_string(&key, "exe"),
                install_path: read_string(&key, "path"),
            };
            out.push(fields);
        }
        out
    }

    fn read_string(key: &RegKey, name: &str) -> String {
        key.get_value::<String, _>(name)
            .unwrap_or_default()
            .trim()
            .to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_entry_requires_name_and_id() {
        let f = GogGameFields {
            game_name: "".into(),
            game_id: "1207".into(),
            exe: "C:/Games/x.exe".into(),
            install_path: "".into(),
        };
        assert!(build_entry(&f).is_none());

        let f = GogGameFields {
            game_name: "Witcher".into(),
            game_id: "".into(),
            exe: "".into(),
            install_path: "".into(),
        };
        assert!(build_entry(&f).is_none());
    }

    #[test]
    fn build_entry_picks_exe_then_path() {
        let f = GogGameFields {
            game_name: "Witcher 3".into(),
            game_id: "1207664663".into(),
            exe: "C:/Games/Witcher3/witcher3.exe".into(),
            install_path: "C:/Games/Witcher3".into(),
        };
        let entry = build_entry(&f).unwrap();
        assert_eq!(entry.id, "gog:1207664663");
        assert_eq!(entry.name, "Witcher 3");
        assert!(matches!(entry.source, AppSource::Gog));
        assert_eq!(
            entry.launch_command,
            "goggalaxy://openGameView/1207664663"
        );
        assert_eq!(
            entry.exe_path.as_ref().unwrap().to_string_lossy(),
            "C:/Games/Witcher3/witcher3.exe"
        );
        assert_eq!(
            entry.metadata.get("game_id").and_then(Value::as_str),
            Some("1207664663")
        );
    }

    #[test]
    fn build_entry_falls_back_to_path_when_exe_empty() {
        let f = GogGameFields {
            game_name: "Stardew".into(),
            game_id: "1453375253".into(),
            exe: "".into(),
            install_path: "C:/GOG Games/Stardew Valley".into(),
        };
        let entry = build_entry(&f).unwrap();
        assert_eq!(
            entry.exe_path.as_ref().unwrap().to_string_lossy(),
            "C:/GOG Games/Stardew Valley"
        );
    }

    #[test]
    fn build_entry_trims_name_whitespace() {
        let f = GogGameFields {
            game_name: "  Cyberpunk 2077  ".into(),
            game_id: "1423049311".into(),
            exe: "".into(),
            install_path: "".into(),
        };
        let entry = build_entry(&f).unwrap();
        assert_eq!(entry.name, "Cyberpunk 2077");
        assert!(entry.exe_path.is_none());
    }

    #[test]
    fn provider_unavailable_offline_when_not_windows() {
        // Smoke: scan() must not panic when registry isn't there.
        let p = GogProvider::new();
        let _ = p.is_available();
        let _ = p.scan();
    }
}
